diff --git a/src/file_descriptor.cpp b/src/file_descriptor.cpp
index c8608c4..d062b43 100644
--- a/src/file_descriptor.cpp
+++ b/src/file_descriptor.cpp
@@ -181,36 +181,27 @@ void file_descriptor_impl::open(const detail::path& p, BOOST_IOS::openmode mode)
         // Calculate oflag argument to open.
 
     int oflag = 0;
-    if ( (mode & (BOOST_IOS::in | BOOST_IOS::out))
-             ==
-         (BOOST_IOS::in | BOOST_IOS::out) )
-    {
-        if( mode & BOOST_IOS::app )
-            boost::throw_exception(BOOST_IOSTREAMS_FAILURE("bad open mode"));
-        oflag |= O_RDWR;
-        if( mode & BOOST_IOS::trunc ) {
-            oflag |= O_TRUNC;
-            oflag |= O_CREAT;
-        }
-    } else if (mode & BOOST_IOS::in) {
-        if( mode & (BOOST_IOS::app | BOOST_IOS::trunc) )
-            boost::throw_exception(BOOST_IOSTREAMS_FAILURE("bad open mode"));
-        oflag |= O_RDONLY;
-    } else if (mode & BOOST_IOS::out) {
-        if( (mode & (BOOST_IOS::app | BOOST_IOS::trunc))
-               ==
-            (BOOST_IOS::app | BOOST_IOS::trunc) )
-            boost::throw_exception(BOOST_IOSTREAMS_FAILURE("bad open mode"));
-        oflag |= O_WRONLY;
-        if (mode & BOOST_IOS::app)
-            oflag |= O_APPEND;
-        else {
-            oflag |= O_CREAT;
-            oflag |= O_TRUNC; 
-        }
-    } else {
+    if ( !(mode & (BOOST_IOS::in | BOOST_IOS::out | BOOST_IOS::app)) ||
+            ((mode & BOOST_IOS::trunc) &&
+            ((mode & BOOST_IOS::app) || !(mode & BOOST_IOS::out))) ) {
         boost::throw_exception(BOOST_IOSTREAMS_FAILURE("bad open mode"));
     }
+    else if ( mode & BOOST_IOS::in ) {
+        if ( mode & BOOST_IOS::app )
+            oflag |= O_CREAT | O_APPEND | O_RDWR;
+        else if ( mode & BOOST_IOS::trunc )
+            oflag |= O_CREAT | O_TRUNC | O_RDWR;
+        else if ( mode & BOOST_IOS::out )
+            oflag |= O_RDWR;
+        else
+            oflag |= O_RDONLY;
+    }
+    else {
+        if ( mode & BOOST_IOS::app )
+            oflag |= O_CREAT | O_APPEND | O_WRONLY;
+        else
+            oflag |= O_CREAT | O_TRUNC | O_WRONLY;
+    }
     #ifdef _LARGEFILE64_SOURCE
         oflag |= O_LARGEFILE;
     #endif
@@ -227,6 +218,12 @@ void file_descriptor_impl::open(const detail::path& p, BOOST_IOS::openmode mode)
     if (fd == -1) {
         boost::throw_exception(system_failure("failed opening file"));
     } else {
+        if ( mode & BOOST_IOS::ate ) {
+            if (BOOST_IOSTREAMS_FD_SEEK(fd, 0, SEEK_END) == -1) {
+                BOOST_IOSTREAMS_FD_CLOSE(fd);
+                boost::throw_exception(system_failure("failed opening file"));
+            }
+        }
         handle_ = fd;
         flags_ = close_always;
     }
diff --git a/src/zlib.cpp b/src/zlib.cpp
index d765e85..56bea9e 100644
--- a/src/zlib.cpp
+++ b/src/zlib.cpp
@@ -76,7 +76,7 @@ void zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(int error)
     switch (error) {
     case Z_OK: 
     case Z_STREAM_END: 
-    //case Z_BUF_ERROR: 
+    case Z_BUF_ERROR: 
         return;
     case Z_MEM_ERROR: 
         boost::throw_exception(std::bad_alloc());
